src/vs/editor/common/model/bracketPairColorizer/ast.ts

```diff
@@ -458,9 +458,11 @@ export class InvalidBracketAstNode extends BaseAstNode {
 
 	public readonly unopenedBrackets: SmallImmutableSet<number>;
 
-	constructor(category: number, length: Length, denseKeyProvider: DenseKeyProvider<number>) {
+	constructor(pairsWith: number[], length: Length, denseKeyProvider: DenseKeyProvider<number>) {
```

Comment:
Have you tried using `category: number` instead?
All this code is highly performance critical, and it would be nice if we don't need an array here.

Basically, if a symbol `begin1` is closed by `end` and `begin2` is also closed by `end`, then `begin1` and `begin2` currently have distinct categories. Even more, `end` is registered twice with both the category of `begin1` and of `begin2`. However, only the first registration of `end` is actually being matched, so `begin2` and `end` never match.

I think a more performant solution would be to assign `begin1` and `begin2` the same category (which makes sense, as they are interchangeable from the perspective of bracket pair colorization). Then everything else would work out of the box.

Reply:
At least, `pairsWith` should also be an `SmallImmutableSet` I think.

Reply:
I agree, re: `SmallImmutableSet`. I can make that change if you'd like.

Reply:
And yes, for the record, my first attempt was as you suggested, and it works for shared brackets which only have one-to-many relationships, but there are issues when brackets have many-to-many relationships. The plugin I'm using to test this scenario is `LaTeX Workshop`, which has brackets defined as follows:
```json
["{", "}"],
["[", "]"],
["(", ")"],
["\\left(", "\\right)"],
["\\left(", "\\right."],
["\\left.", "\\right)"],
["\\left[", "\\right]"],
["\\left[", "\\right."],
["\\left.", "\\right]"],
["\\left\\{", "\\right\\}"],
["\\left\\{", "\\right."],
["\\left.", "\\right\\}"],
["\\left<", "\\right>"]
```
Trying, for example, to get `\left( \right.`, `\left. \right)`, but not `\left. \right.` to match is what made me jump to using a `pairsWith` list.

Reply:
I think there should be no performance impact for languages that just have `{` ... `}`, `(` ... `)` and `[`...`]`.

Probably `SmallImmutableSet` can do that.